TripalCultivate/TripalCultivate-Phenotypes
# G4.87 revise genus validator

Issue #87

## Motivation
This draft PR is an installment of the genus+project issue. Currently this draft proposes a genus validator plugin and project validator will follow.

**Update 07/09: projectExists Validator**
project exists validator will check project provided exists in chado.projects table.

**Update 07/09: projectGenusMatch Validator**
project genus match validator will check project provided exists in chado.projects table and inspect the genus set to project matches the genus provided.




## What does this PR do?
*Please describe each things this PR does. For example, a PR may 1) solve a specific bug, 2) create an auomated test to ensure it doesn't return.*

1. Check that genus provided in the importer exists
2. Check that genus provided is configured.

**Update 07/09: projectExists Validator**
1. Check that a project id or project name exists in chado.projects table.

**Update 07/11: ProjectGenusMatch Validator**
1. Check that a project with genus set match the value provided in the genus field.

## Testing

### Automated Testing
*Please describe each automated test this PR creates and provide a list of the assertions it makes using casual language.*
*Do not just say things like "asserts the array is not empty" but rather say "Ensures that the return value of method X with these parameters is not an empty array".*

Test genus input.
1. Test that an exception is thrown when:
- the $form_state was passed as parameter to the validator instead of doing $form_state->getValues()
- No field with the name/key 'genus' detected in the $form_state

2. Test non-existent genus
4. Test an existing genus but is not configured.
5. Test an existing genus and is configured.

**Update 07/09: projectExists Validator**
Test project input.
1. Test that an exception is thrown when:
- the $form_state was passed as parameter to the validator instead of doing $form_state->getValues()
-  A string value.
- No field with the name/key 'project' detected in the $form_state

2. Test non-existent project
4. Test an existing project using project id and project name.

**Update 07/11: projectGenusMatch Validator**
Test project and genus input.
1. Test that an exception is thrown when:
- the $form_state was passed as parameter to the validator instead of doing $form_state->getValues()
-  A string value.
- No field with the name/key 'project' or 'genus' detected in the $form_state

2. Test non-existent project
4. Test an existing project with a genus set matches  the genus provided in the genus field.

### Manual Testing
*Describe in detail how someone should manually test this functionality.*
*Make sure to include whether they need to build a docker from scratch, create any records, etc.*

**Update 07/09: projectExists Validator**
1. Create a project project record.
2. In your D10/T4 instance, access my site/devel/php
```
$plugin_manager = \Drupal::service('plugin.manager.trpcultivate_validator');
$validator_id = 'trpcultivate_phenotypes_validator_project_exists';
$instance = $plugin_manager->createInstance($validator_id);

// PASS
$form_values = ['project' => 'REPLACE WITH PROJECT ID OR PROJECT NAME'];
$validation_status = $instance->validateMetadata($form_values);
print_r($validation_status);

// FAIL: not found.
$form_values = ['project' => 'out_of_this_world_project'];
$validation_status = $instance->validateMetadata($form_values);
print_r($validation_status);

// Will throw an exception - execute separately
$form_values = 'A string value';
$validation_status = $instance->validateMetadata($form_values);

$form_values = ['genus' => 'Lens', 'file' => 2];
$validation_status = $instance->validateMetadata($form_values);

use Drupal\Core\Form\FormState;
$form_state = new FormState();
$form_state->setValues(['project' => 'abc']);
$validation_status = $instance->validateMetadata($form_state);
```

**Update 07/11: projectGenusMatch Validator**
1. Create a project project record.
2. Create and configure a genus.
3. Setup a project-genus property

```
\Drupal::service('trpcultivate_phenotypes.genus_project')
  ->setGenusToProject(REPLACE WITH PROJECT ID, REPLACE WITH CONFIGURED GENUS);
```

4. In your D10/T4 instance, access my site/devel/php
```
$plugin_manager = \Drupal::service('plugin.manager.trpcultivate_validator');
$validator_id = 'trpcultivate_phenotypes_validator_project_genus_match';
$instance = $plugin_manager->createInstance($validator_id);

// PASS
$form_values = ['project' => 'REPLACE WITH PROJECT ID OR PROJECT NAME', 'genus' => 'REPLACE WITH CONFIGURED GENUS'];
$validation_status = $instance->validateMetadata($form_values);
print_r($validation_status);

// FAIL: not found.
$form_values = ['project' => 'out_of_this_world_project', 'genus' => 'REPLACE WITH CONFIGURED GENUS'];
$validation_status = $instance->validateMetadata($form_values);
print_r($validation_status);

// FAIL: not the correct genus.
$form_values = ['project' => 'REPLACE WITH PROJECT ID', 'genus' => 'Not the genus'];
$validation_status = $instance->validateMetadata($form_values);
print_r($validation_status);

// Will throw an exception - execute separately
$form_values = 'A string value';
$validation_status = $instance->validateMetadata($form_values);

// Missing project field.
$form_values = ['genus' => 'Lens', 'file' => 2];
$validation_status = $instance->validateMetadata($form_values);

// Missing genus field.
$form_values = ['project' => 'The project', 'file' => 2];
$validation_status = $instance->validateMetadata($form_values);

use Drupal\Core\Form\FormState;
$form_state = new FormState();
$form_state->setValues(['project' => 'abc']);
$validation_status = $instance->validateMetadata($form_state);
```

//////////////////////////////////////////////


1. Setup 2 genus where 1 is configured whereas the other is not configured.
2. In your D10/T4 instance, access my site/devel/php


```
$plugin_manager = \Drupal::service('plugin.manager.trpcultivate_validator');
$validator_id = 'trpcultivate_phenotypes_validator_genus_exists';
$instance = $plugin_manager->createInstance($validator_id);

// PASS
$form_values = ['genus' => 'REPLACE WITH CONFIGURED GENUS'];
$validation_status = $instance->validateMetadata($form_values);
print_r($validation_status);


// ------------------------------


// FAIL: not configured.
$form_values = ['genus' => 'REPLACE WITH NOT CONFIGURED GENUS'];
$validation_status = $instance->validateMetadata($form_values);
print_r($validation_status);


// ------------------------------

// FAIL: not found.
$form_values = ['genus' => 'spurious_genus'];
$validation_status = $instance->validateMetadata($form_values);
print_r($validation_status);

// -----------------------------

// Will throw an exception - execute separately
$form_values = ['project' => 1, 'file' => 2];
$validation_status = $instance->validateMetadata($form_values);

use Drupal\Core\Form\FormState;
$form_state = new FormState();
$form_state->setValues(['genus' => 'abc']);
$validation_status = $instance->validateMetadata($form_state);
```
